// AGL/user/factory.js
/*
 *  @file Creates user objects. These objects are immutable, so it can be
 *  assumed that all user object are valid. (Validation is done on creation)
 *  @author Matt Lambert
 */

'use strict';
//==============================================================================
//  Modules
//==============================================================================

var builder = require('AGL/builder');

//==============================================================================
//  Constants
//==============================================================================

//  User property map - See AGL/builder module for correct structure
var USER_PROPERTIES = {
	publicId     : {
		required         : true,
		type             : 'AGLPublicId'
	},
	username     : {
		required         : true,
		type             : 'AGLUsername'
	},
	passwordHash : {
		required         : true,
		type             : 'AGLPasswordHash'
	},
	email        : {
		required         : true,
		type             : 'email'
	},
	created      : {
		required         : false,
		type             : 'datetime',
		database_default : true
	},
};

var USER_BUILDER = builder.NewBuilder(USER_PROPERTIES, 'User');

//==============================================================================
//  Public functions
//==============================================================================

/*  public object BuildUser()
 *
 *  Returns a user builder used to build a new user object.
 *
 *  @return {object} USER_BUILDER
 */
module.exports.BuildUser = function () {
	return USER_BUILDER;
};
