// AGL/user/repository.js
/*
 *   @file Responsible for saving, updating, and deleting user objects to/from the database.
 *   @author Matt Lambert
 */

'use strict';
//==============================================================================
//  Modules
//==============================================================================

var dbAccessor  = require('AGL/database/dbAccessor');
var userFactory = require('AGL/user/factory');
var validate    = require('validate.js');

//==============================================================================
//  Constants
//==============================================================================

var TABLE_NAME = 'users';

//==============================================================================
//  Public Functions
//==============================================================================

/* 
 *  public undefined SaveUser(user, callback)
 *
 *  Saves a new or existing user to the database.
 *  Returns the user's username to the callback if save is successfull.
 *
 *  @param {object} user object, generated from a factory.
 *  @param {function} callback
 */
module.exports.SaveUser = function (user, callback) {
	if (!validate.isFunction(callback)) {
		throw new Error('callback argument must be a function.');
	}

	// TODO When I change this to either insert or update a user, use async to avoid nested callback hell

	InsertUser(user, function(error, rows) {
		if (error) {
			return callback(error);
		}

		return callback(null, rows[0]);// TODO need to actually return username
	});

};

/* public undefined GetUser(username, callback)
 *
 *  Gets a user from the database based on username.
 *  Returns a user object to the callback if succussfull.
 *
 *  @param {string} username
 *  @param {function} callback
 */
module.exports.GetUser = function (username, callback) {
	if (!validate.isFunction(callback)) {
		throw new Error('callback argument must be a function.');
	}

};

//==============================================================================
//  Private Functions
//==============================================================================

// TODO DOCO
function QueryUser (username, callback) {
	dbAccessor.GetQueryBuilder()(TABLE_NAME)
	.select('username', 'password', 'email', 'created', 'lastModified')
	.where({
		username : username
	})
	.limit(1)
	.asCallback(callback);
}

// TODO DOCO
function InsertUser (user, callback) {
	dbAccessor.GetQueryBuilder()(TABLE_NAME)
	.insert({
		username : user.GetUsername(),
		password : user.GetPasswordHash(),
		email    : user.GetEmail()
	})
	.asCallback(callback);
}

// TODO DOCO
function UpdateUser (user, callback) {
	dbAccessor.GetQueryBuilder()(TABLE_NAME)
	.update({
		username : user.GetUsername(),
		password : user.GetPasswordHash(),
		email    : user.GetEmail()
	})
	.where({
		username : user.GetUsername(),
	})
	.asCallback(callback);
}

// TODO DOCO
function DeleteUser (username, callback) {
	dbAccessor.GetQueryBuilder()(TABLE_NAME)
	.del()
	.where({
		username : username,
	})
}
