// AGL/routes.js
'use strict';
//==============================================================================
//  Modules
//==============================================================================

var userAPI     = require('AGL/api/user');

//==============================================================================
//  Exports
//==============================================================================

module.exports = function (app) {

    // API routes

    UserAPIRoutes(app);

    // Front end routes

    app.get('*', function (request, response) {
        response.sendFile('index.html', {'root' : './public'});
    });
};

//==============================================================================
//  Handle Response
//==============================================================================

function HandleResponse (error, data, response) {
    var APIResponse;
    if (error) {
        console.log("ERROR: " + error);
        APIResponse = error;
    }
    else {
        console.log("DATA: ");
        console.log(data);
        APIResponse = data;
    }
    response.json(APIResponse);
}

//==============================================================================
//  User API
//==============================================================================

function UserAPIRoutes(app) {

    app.post('/api/users', function (request, response) {

        userAPI.CreateUser({
            public_id   : request.body.public_id,
            username    : request.body.username,
            password    : request.body.password,
            email       : request.body.email,
        },
        function UserCreated (error, data) {
            HandleResponse(error, data, response);
        });

    });

    app.get('/api/users/:public_id', function (request, response) {

        userAPI.GetUser({
            public_id   : request.params.public_id,
        },
        function UserFound (error, data) {
            HandleResponse(error, data, response);
        });
    });

}
