// AGL/mock/database/dbAccessorMock.js
/*
 *   @file Mocks dbAccessor for testing purposes.
 *   @author Matt Lambert
 */

// TODO This is super messy - needs tidyup

'use strict';
//==============================================================================
//  Public Functions
//==============================================================================

module.exports.GetQueryBuilder = function () {
	return KnexMockFunction;
};

module.exports.SetCallback = function (callback) {
	ResultsCallback = callback;
};

//==============================================================================
//  Private Functions
//==============================================================================

var ResultsCallback = function (callback) {
	return callback(null, []);
}

var KnexMock = function (table) {
	this.insert = function () { return this; };
	this.select = function () { return this; };
	this.where = function () { return this; };
	this.orWhere = function () { return this; };
	this.update = function () { return this; };
	this.limit = function () { return this; };
	this.del = function () { return this; };
	this.asCallback = ResultsCallback;
};

var KnexMockFunction = function (table) {
    return new KnexMock(table);
};
