// AGL/database/db_accessor.js
"use strict";
//==============================================================================
//  Modules
//==============================================================================

var config  = require('config');
var mysql   = require('mysql');

//==============================================================================
//  Public functions
//==============================================================================

//  Executes the provided query and calls the callback
//  with the results or an error.
//
//  @param {string} query
//  @param {array} parameters
//  @param {function} callback
//
module.exports.ExecuteQuery = function(query, parameters, callback) {
    (typeof callback !== 'function') ? throw("Callback argument must be a function.");

    var connnection = getConnection();
    connection.connect();
    connection.query(sql, parameters, function queryComplete(error, results, fields) {
        if (error) {
            callback(error, null);
        }
        else {
            callback(null, results);
        }
    });
    connection.end();
};

//==============================================================================
//  Private functions
//==============================================================================


//  Gets a connection to the database.
//
//  @return {object} database connection
//
function GetConnection() {
    var connection = mysql.createConnection({
        host        : config.db.host,
        user        : config.db.user,
        password    : config.db.password,
        database    : config.db.database
    });
    return connection;
}
