// AGL/database/db_accessor.js
'use strict';
//==============================================================================
//  Modules
//==============================================================================

var config  = require('AGL/config');
console.log('Initilising knex with config');
console.log(config);
var knex   = require('knex')({
    debug       : config.development,
    client      : config.db.client,
    connection  : {
        host        : config.db.host,
        user        : config.db.user,
        password    : config.db.password,
        database    : config.db.database,
    },
});

//==============================================================================
//  Public functions
//==============================================================================

//  public object GetQueryBuilder()
//
//  Returns the knex query builder object so
//  queries can be performed against the DB.
//
//  @return {object} knex query builder object
//
module.exports.GetQueryBuilder = function () {
    return knex;
};
