// AGL/database/dbAccessor.js
/*
 *  @file Provides access to the database via a query builder. 
 *  @author Matt Lambert
 */

'use strict';
//==============================================================================
//  Modules
//==============================================================================

var validate       = require('validate.js');
var databaseConfig = require('AGL/AGLConfig').LoadConfig('database');
var knex           = require('knex')({
    client      : databaseConfig.client,
    connection  : {
        socketPath  : databaseConfig.socketPath,
        host        : databaseConfig.host,
        user        : databaseConfig.user,
        password    : databaseConfig.password,
        database    : databaseConfig.database,
    },
});

//==============================================================================
//  Public functions
//==============================================================================

/*
 *  public object GetQueryBuilder()
 *
 *  Returns the query builder used to execute SQL queries
 *
 *  @return {object} query builder
 */
module.exports.GetQueryBuilder = function () {
	return knex;
};
